tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,429 @@
+
+            if (!server_cert) {
+                err_code = S2N_CERT_ERR_INVALID;
+                goto clean_up;
+            }
+
+            //add the cert to the chain.
+            if (!sk_X509_push(validator->cert_chain, server_cert)) {
+                X509_free(server_cert);
+                err_code = S2N_CERT_ERR_INVALID;
+                goto clean_up;
+            }
+        }
+
+        /* Pull the public key from the first certificate */
+        if (certificate_count == 0) {
+            /* Assume that the asn1cert is an RSA Cert */
+            if (s2n_asn1der_to_public_key(&public_key_out->pkey, &asn1cert) < 0) {
+                err_code = S2N_CERT_ERR_INVALID;
+                goto clean_up;
+            }
+            if (s2n_cert_public_key_set_cert_type(public_key_out, S2N_CERT_TYPE_RSA_SIGN) < 0) {
+                err_code = S2N_CERT_ERR_INVALID;
+                goto clean_up;
+            }
+
+            public_key = server_cert;
+        }
+
+        last_in_chain = server_cert;
+
+        certificate_count++;
+    }
+
+    if (certificate_count < 1) {
+        err_code = S2N_CERT_ERR_INVALID;
+        goto clean_up;
+    }
+
+    if (validator->validate_certificates && validator->verify_host_fn &&
+        !verify_host_information(validator, public_key)) {
+        err_code = S2N_CERT_ERR_UNTRUSTED;
+        goto clean_up;
+    }
+
+    if (validator->validate_certificates) {
+        //now that we have a chain, get the store and check against it.
+        ctx = X509_STORE_CTX_new();
+        int op_code = X509_STORE_CTX_init(ctx, validator->trust_store->trust_store, last_in_chain,
+                                          validator->cert_chain);
+
+        if (op_code <= 0) {
+            err_code = S2N_CERT_ERR_INVALID;
+            goto clean_up;
+        }
+
+        op_code = X509_verify_cert(ctx);
+
+        if (op_code <= 0) {
+            op_code = X509_STORE_CTX_get_error(ctx);
+
+            X509_STORE_CTX_cleanup(ctx);
+            err_code = S2N_CERT_ERR_UNTRUSTED;
+            goto clean_up;
+        }
+    }
+
+    clean_up:
+    if (ctx) {
+        X509_STORE_CTX_cleanup(ctx);
+    }
+    s2n_stuffer_free(&cert_chain_in_stuffer);
+    return err_code;
+}
+
+
+s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(struct s2n_x509_validator *validator,
+                                                                                const uint8_t *ocsp_response_raw,
+                                                                                uint32_t ocsp_response_length,
+                                                                                struct s2n_config *config) {
+    if (!validator->validate_certificates || !validator->check_stapled_ocsp) {
+        return S2N_CERT_OK;
+    }
+
+    OCSP_RESPONSE *ocsp_response = NULL;
+    OCSP_BASICRESP *basic_response = NULL;
+
+    s2n_cert_validation_code ret_val = S2N_CERT_OK;
```

Comment:
Same comment, default should be `S2N_CERT_ERR_INVALID`.

Reply:
fixed